### PR TITLE
Develop remove past action communication

### PR DESCRIPTION
**Pull request by vincentpierre**

### CODE REVIEW

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/GrpcExtensions.cs [hunk 1]
```diff
@@ -11,15 +11,54 @@ namespace MLAgents
...
+                agentInfoProto.ActionMask.AddRange(ai.actionMasks);
+            }
+
+            if (ai.observations != null)
+            {
+                foreach (var obs in ai.observations)
+                {
+                    agentInfoProto.Observations.Add(obs.ToProto());
+                }
+            }
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Grpc/GrpcExtensions.cs]: Can't this block just be `ai.ToAgentInfoProto()`?

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs [hunk 2]
```diff
@@ -198,8 +207,18 @@ public void Generate(TensorProxy tensorProxy, int batchSize, IEnumerable<Agent>
             {
                 var offset = memorySize * m_MemoryIndex;
                 List<float> memory;
+                //Debug.Log("Memories used");
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs]: nit: remove commented out code.

File: docs/Python-API.md [hunk 3]
```diff
@@ -85,9 +85,6 @@ A BrainInfo object contains the following fields:
 - **`max_reached`** : A list as long as the number of Agents using the Brain
   containing true if the Agents reached their max steps.
 - **`agents`** : A list of the unique ids of the Agents using the Brain.
-- **`previous_actions`** : A two dimensional numpy array of dimension `(batch
```
Comment (chriselion) [inline docs/Python-API.md]: nit: should also remove `memories` line above this.

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/GrpcExtensions.cs (hunk 1 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/Grpc/GrpcExtensions.cs]: Nice catch

File: ml-agents/mlagents/trainers/tf_policy.py [hunk 4]
```diff
@@ -172,7 +178,7 @@ def make_empty_memory(self, num_agents):
         :param num_agents: Number of agents.
         :return: Numpy array of zeros.
         """
-        return np.zeros((num_agents, self.m_size))
+        return np.zeros((num_agents, self.m_size), dtype=np.float)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/tf_policy.py]: Out of scope for this PR, but should we try to make sure we never create float64 nparrays?

File: ml-agents/mlagents/trainers/tf_policy.py [hunk 5]
```diff
@@ -194,6 +200,34 @@ def remove_memories(self, agent_ids):
             if agent_id in self.memory_dict:
                 self.memory_dict.pop(agent_id)
 
+    def make_empty_previous_action(self, num_agents):
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/tf_policy.py]: Is there any difference between how we handle memories and previous actions besides the the dtype? Worth combining into a small utility class?

**Comment by chriselion** [comment]
Just to clarify, existing models will keep working as before? The only thing that would break when upgrading is existing demo files?

**Comment by vincentpierre** [comment]
Models will still work as before but the demo files will not.

File: ml-agents/mlagents/trainers/tf_policy.py (hunk 5 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/tf_policy.py]: I think there is going to be a bigger refactor of this part coming later. I think it is fine for now. What do you think @ervteng ?

**Comment by ervteng** [comment]
> Models will still work as before but the demo files will not.

Will we need to re-record all of them or just the Hallway one?

**Comment by vincentpierre** [comment]
> Will we need to re-record all of them or just the Hallway one?

I reconverted them already

File: ml-agents/mlagents/trainers/tf_policy.py (hunk 5 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tf_policy.py]: Might be worth combining - but then again, there will only be two things we store between steps. Also, the "when" they're added is different - memories can be added after the get_action, previous_actions must be added before. 

I think there is no way around it, this logic will still need to live in the Policy for Python inference to work. The only other place I can think of is the env_manager, but that would be messy.

**Pull request merged**